sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala

```diff
@@ -254,22 +254,28 @@ private[hive] trait HiveStrategies {
    * Retrieves data using a HiveTableScan.  Partition pruning predicates are also detected and
    * applied.
    */
-  object HiveTableScans extends Strategy {
+  object HiveTableScans extends Strategy with PredicateHelper {
     def apply(plan: LogicalPlan): Seq[SparkPlan] = plan match {
-      case ScanOperation(projectList, predicates, relation: HiveTableRelation) =>
+      case ScanOperation(projectList, filters, relation: HiveTableRelation) =>
         // Filter out all predicates that only deal with partition keys, these are given to the
         // hive table scan operator to be used for partition pruning.
         val partitionKeyIds = AttributeSet(relation.partitionCols)
-        val (pruningPredicates, otherPredicates) = predicates.partition { predicate =>
-          !predicate.references.isEmpty &&
-          predicate.references.subsetOf(partitionKeyIds)
+        val normalized = DataSourceStrategy.normalizeExprs(
+          filters.filter(_.deterministic), relation.output)
+        val partitionKeyFilters = if (relation.partitionCols.isEmpty) {
+          ExpressionSet(Nil)
+        } else {
+          val predicates = ExpressionSet(normalized
+            .flatMap(extractPredicatesWithinOutputSet(_, partitionKeyIds)))
+          logInfo(s"Pruning partitions with: ${filters.mkString(",")}")
+          predicates
         }
 
         pruneFilterProject(
           projectList,
-          otherPredicates,
+          (ExpressionSet(filters) -- partitionKeyFilters).toSeq,
```

Comment:
Shall we use normalized filters here?

Reply:
Hmm, looks should be fine as `ExpressionSet` will canonicalize expressions.

Reply:
does it really work? `DataSourceStrategy.getPushedDownFilters` may do CNF and return predicates that not in `filters`.

Reply:
@cloud-fan sorry, missed this, you're right, just changed